### PR TITLE
Added a property to the Spring generator to avoid the use of the `ResponseEntity` type

### PR DESCRIPTION
**Pull request by rcardin**
The PR adds the boolean property `useResponseEntity` to the Spring generator. Through the property, it's to generate controllers and APIs that return directly a value, avoiding its wrapping inside a `ResponseEntity`.

The PR fixed issue #6135.

**I cannot understand how to add automatic tests verifying the generated files. Please, help me with some hints!**

As requested, I tag the Java language technical committee: @bbdouglas, @sreeshas, @jfiala, @lukoyanov, @cbornet , @jeff9finger, @karismann, @Zomzog, @lwlee2608

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the `master` branch
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by wing328** [comment]
@rcardin can you please resolve the merge conflicts when you've time? Thanks.

**Comment by rcardin** [comment]
> @rcardin can you please resolve the merge conflicts when you've time? Thanks.

@wing328, sure thing! Done.

**Comment by rcardin** [comment]
@wing328, I'm sorry. The last commit I made is a mess :(
I can close this PR and open a new one if you prefer.

Just let me know.

**Comment by welshm** [comment]
Thanks for the PR! Taking a look...

So your intent is that when generating the interface, you don't want to use `ResponseEntity<>` provided by Spring and instead just want to return the type? Can you elaborate on _why_ you would want to do that when `ResponseEntity<>` allows you more control over the response?

To add a test, you can look at `SpringCodegenTest.java` - specifically `doGeneratePathVariableForSimpleParam`. You can use an existing `yaml` in your case to validate that response entities are not made (although I recommend picking one that has both void and other types of returns).

If you have any questions, feel free to reply to me here or reach out to me on the [Slack](https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g)!

**Comment by rcardin** [comment]
Hi, @welshm, thanks for the response.
> Can you elaborate on why you would want to do that when ResponseEntity<> allows you more control over the response?

The use of controller advices and Spring annotations over the methods implementing HTTP endpoints leave very few cases in which a `ResponseEntity<>` adds a real value. IMHO, a plain method returning pure DTOs is more readable and maintainable than one that returns a `ResponseEntity<>`.

I will check the tests examples you pointed me as soon as possible.

Cheers.

**Comment by borsch** [comment]
@rcardin recently I've added few classes that can be used to assert generated files content. You can check this MR for reference - https://github.com/OpenAPITools/openapi-generator/pull/11738

File: modules/openapi-generator/src/main/resources/JavaSpring/methodBody.mustache [hunk 1]
```diff
@@ -14,14 +14,16 @@ return CompletableFuture.supplyAsync(()-> {
     {{#-last}}
 {{#async}}        {{/async}}{{^async}}    {{/async}}        }
 {{#async}}    {{/async}}        });
-{{#async}}    {{/async}}        return new ResponseEntity<>({{#returnSuccessCode}}HttpStatus.valueOf({{{statusCode}}}){{/returnSuccessCode}}{{^returnSuccessCode}}HttpStatus.NOT_IMPLEMENTED{{/returnSuccessCode}});
+{{#async}}    {{/async}}        {{#useResponseEntity}}return new ResponseEntity<>({{#returnSuccessCode}}HttpStatus.valueOf({{{statusCode}}}){{/returnSuccessCode}}{{^returnSuccessCode}}HttpStatus.NOT_IMPLEMENTED{{/returnSuccessCode}});{{/useResponseEntity}}
+{{^useResponseEntity}}        throw new IllegalArgumentException("Not implemented");{{/useResponseEntity}}
```
Comment (borsch) [inline modules/openapi-generator/src/main/resources/JavaSpring/methodBody.mustache:18]: don't add new line here

File: modules/openapi-generator/src/main/resources/JavaSpring/methodBody.mustache [hunk 2]
```diff
@@ -14,14 +14,16 @@ return CompletableFuture.supplyAsync(()-> {
...
             {{#async}}
         }, Runnable::run);
             {{/async}}
     {{/-last}}
 {{/examples}}
 {{^examples}}
-return {{#async}}CompletableFuture.completedFuture({{/async}}new ResponseEntity<>({{#returnSuccessCode}}HttpStatus.OK{{/returnSuccessCode}}{{^returnSuccessCode}}HttpStatus.NOT_IMPLEMENTED{{/returnSuccessCode}}){{#async}}){{/async}};
+{{#useResponseEntity}}return {{#async}}CompletableFuture.completedFuture({{/async}}new ResponseEntity<>({{#returnSuccessCode}}HttpStatus.OK{{/returnSuccessCode}}{{^returnSuccessCode}}HttpStatus.NOT_IMPLEMENTED{{/returnSuccessCode}}){{#async}}){{/async}};{{/useResponseEntity}}
+{{^useResponseEntity}}return {{#jdk8}}{{#async}}CompletableFuture.runAsync(() ->{{/async}}{{/jdk8}}throw new IllegalArgumentException("Not implemented"){{#jdk8}}{{#async}}){{/async}}{{/jdk8}};{{/useResponseEntity}}
```
Comment (borsch) [inline modules/openapi-generator/src/main/resources/JavaSpring/methodBody.mustache:26]: here as well

**Review by borsch — CHANGES_REQUESTED** [review]

**Comment by anatoliy-balakirev** [comment]
Hi @welshm. 

> Can you elaborate on why you would want to do that when ResponseEntity<> allows you more control over the response?

Another benefit of using the suggested approach is that the response status may be generated instead of being added manually. For example, if my endpoint is supposed to return `void` - the response code may be automatically generated like this:
```
@ResponseStatus(HttpStatus.NO_CONTENT)
```
Instead of manually adding this to the code:
```
return ResponseEntity.noContent().build();
```
In fact, I saw many situations, where the spec was defining 200 instead of 204 for such cases, but people were mechanically adding `return ResponseEntity.noContent().build();` as that seemed to be logical for the `void` return type. Ending up with the impl not matching the spec.

**Comment by borsch** [comment]
@rcardin could you please sync this MR with latest master?

**Comment by rcardin** [comment]
Ok, I'll try during this week.

**Comment by borsch** [comment]
@rcardin also please make sure that this will work with `delegatePattern: "true"`. Example in `bin/configs/spring-boot-delegate-j8.yaml`

**Comment by borsch** [comment]
@rcardin just noticed your query on unit tests. You can check this example
https://github.com/OpenAPITools/openapi-generator/blob/6d48d7342112f3a4d1c543054575a21424ec8f45/modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java#L86

**Comment by rcardin** [comment]
@borsch , in your opinion, which is the right configuration for testing the `delegationPattern` variable set to `true`. I'm trying the following:
```java
Map<String, Object> additionalProperties = new HashMap<>();
additionalProperties.put(AbstractJavaCodegen.FULL_JAVA_UTIL, "true");
additionalProperties.put(SpringCodegen.USE_TAGS, "true");
additionalProperties.put(SpringCodegen.INTERFACE_ONLY, "true");
additionalProperties.put(SpringCodegen.SKIP_DEFAULT_INTERFACE, "true");
additionalProperties.put(SpringCodegen.PERFORM_BEANVALIDATION, "true");
additionalProperties.put(SpringCodegen.SPRING_CONTROLLER, "true");
additionalProperties.put(CodegenConstants.SERIALIZATION_LIBRARY, "jackson");
additionalProperties.put(USE_RESPONSE_ENTITY, "false");
additionalProperties.put(DELEGATE_PATTERN, "true");
```
However, it doesn't generate any delegate class, mostly due to the `SpringCodegen.INTERFACE_ONLY`property set to `true`:/

**Comment by borsch** [comment]
@rcardin yeah, you have to remove `SpringCodegen.INTERFACE_ONLY` from configuration

**Comment by rcardin** [comment]
@borsch, I think I'm done.

File: modules/openapi-generator/src/main/resources/JavaSpring/api.mustache [hunk 3]
```diff
@@ -233,7 +235,7 @@ public interface {{classname}} {
...
+    {{#jdk8-default-interface}}default {{/jdk8-default-interface}} {{#responseWrapper}}{{.}}<{{/responseWrapper}}{{#useResponseEntity}}ResponseEntity<{{/useResponseEntity}}{{>returnTypes}}{{#useResponseEntity}}>{{/useResponseEntity}}{{#responseWrapper}}>{{/responseWrapper}} {{operationId}}({{#allParams}}{{^isFile}}{{^isBodyParam}}{{>optionalDataType}}{{/isBodyParam}}{{#isBodyParam}}{{^reactive}}{{{dataType}}}{{/reactive}}{{#reactive}}{{^isArray}}Mono<{{{dataType}}}>{{/isArray}}{{#isArray}}Flux<{{{baseType}}}>{{/isArray}}{{/reactive}}{{/isBodyParam}}{{/isFile}}{{#isFile}}{{#reactive}}Flux<Part>{{/reactive}}{{^reactive}}MultipartFile{{/reactive}}{{/isFile}} {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}{{#reactive}}{{#hasParams}}, {{/hasParams}}{{#springFoxDocumentationProvider}}@ApiIgnore{{/springFoxDocumentationProvider}} final ServerWebExchange exchange{{/reactive}}{{#vendorExtensions.x-spring-paginated}}, {{#springFoxDocumentationProvider}}@ApiIgnore{{/springFoxDocumentationProvider}} final Pageable pageable{{/vendorExtensions.x-spring-paginated}}){{#unhandledException}} throws Exception{{/unhandledException}} {
```
Comment (anatoliy-balakirev) [inline modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:238]: I may be missing something and sorry if that's the case, but as I understand, if we do not use `ResponseEntity`, then the only way to provide an HTTP status code back (for successful flow) is to add the following annotation to the method: `@ResponseStatus(HttpStatus.<needed status>)`. Will that be autogenerated based on the openapi spec with this change?

Comment (anatoliy-balakirev) [reply modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:238]: I mean, normally if the response is `void` it will probably automatically return 201, and if the response is some object - 200 (I'm not 100% sure about that though). But what if the spec contains something else for whatever reason? Modifying the spec may not be an option and falling back to `ResponseEntity` just for that is a pity... But it probably should not block the current PR anyway.

Comment (rcardin) [reply modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:238]: Maybe, we can add something that reads the returning status from the specification? However, I don't know if it's the main focus of this issue.

**Comment by rcardin** [comment]
@borsch, I'm finalizing the change that added a @ResponseStatus on the method. Just few days and it'll be finished.

**Comment by rcardin** [comment]
@borsch, I pushed the changes. Let me know if everything is ok.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaCamelServerCodegen.java [hunk 4]
```diff
@@ -171,7 +171,7 @@ public void addOperationToGroup(String tag, String resourcePath, Operation opera
     }
 
     private void addCliOptions() {
-        cliOptions.add(new CliOption(CAMEL_REST_COMPONENT, "name of the Camel component to use as the REST consumer").defaultValue(camelRestComponent));
+        cliOptions.add(new CliOption(CAMEL_REST_COMPONENT, "name of the Camel component to u as the REST consumer").defaultValue(camelRestComponent));
```
Comment (borsch) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaCamelServerCodegen.java:174]: unnecessary change

File: modules/openapi-generator/src/main/resources/JavaSpring/api.mustache [hunk 5]
```diff
@@ -222,7 +227,8 @@ public interface {{classname}} {
         headers = { {{#vendorExtensions.versionHeaderParamsList}}"{{baseName}}{{#defaultValue}}={{{.}}}{{/defaultValue}}"{{^-last}}, {{/-last}}{{/vendorExtensions.versionHeaderParamsList}} } {{/hasVersionHeaders}}{{#hasVersionQueryParams}},
         params = { {{#vendorExtensions.versionQueryParamsList}}"{{baseName}}{{#defaultValue}}={{{.}}}{{/defaultValue}}"{{^-last}}, {{/-last}}{{/vendorExtensions.versionQueryParamsList}} } {{/hasVersionQueryParams}}
     )
-    {{#jdk8-default-interface}}default {{/jdk8-default-interface}}{{#responseWrapper}}{{.}}<{{/responseWrapper}}ResponseEntity<{{>returnTypes}}>{{#responseWrapper}}>{{/responseWrapper}} {{#delegate-method}}_{{/delegate-method}}{{operationId}}(
+    @ResponseStatus({{#springHttpStatus}}{{#responses.0}}{{{code}}}{{/responses.0}}{{/springHttpStatus}})
```
Comment (borsch) [inline modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:230]: add this only if `useResponseEntity=false`

File: modules/openapi-generator/src/main/resources/JavaSpring/apiController.mustache [hunk 6]
```diff
@@ -109,7 +111,10 @@ public class {{classname}}Controller implements {{classname}} {
     {{#isDeprecated}}
     @Deprecated
     {{/isDeprecated}}
-    public {{#responseWrapper}}{{.}}<{{/responseWrapper}}ResponseEntity<{{>returnTypes}}>{{#responseWrapper}}>{{/responseWrapper}} {{operationId}}(
+    {{^useResponseEntity}}
+    @ResponseStatus({{#springHttpStatus}}{{{code}}}{{/springHttpStatus}})
```
Comment (borsch) [inline modules/openapi-generator/src/main/resources/JavaSpring/apiController.mustache:115]: not required here since it's present in interface

Comment (borsch) [reply modules/openapi-generator/src/main/resources/JavaSpring/apiController.mustache:115]: bellow there is one more place where `ResponseEntity` is used (Can't put a comment at that line)

File: samples/client/petstore/spring-stubs/src/main/java/org/openapitools/api/PetApi.java [hunk 7]
```diff
@@ -64,6 +64,7 @@ default ResponseEntity<Void> addPet(
     ) {
         return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
 
+
```
Comment (borsch) [inline samples/client/petstore/spring-stubs/src/main/java/org/openapitools/api/PetApi.java:67]: cant' figure out where this new line come from 🥲

**Review by borsch — CHANGES_REQUESTED** [review]
Thanks @rcardin. I left few comments. Overall looks good

File: modules/openapi-generator/src/main/resources/JavaSpring/api.mustache (hunk 5 above)
Comment (rcardin) [reply modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:230]: Nice catch.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaCamelServerCodegen.java (hunk 4 above)
Comment (rcardin) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaCamelServerCodegen.java:174]: I can't figure out where the change comes from

File: modules/openapi-generator/src/main/resources/JavaSpring/apiController.mustache (hunk 6 above)
Comment (rcardin) [reply modules/openapi-generator/src/main/resources/JavaSpring/apiController.mustache:115]: Oh, you're right.

File: samples/client/petstore/spring-stubs/src/main/java/org/openapitools/api/PetApi.java (hunk 7 above)
Comment (rcardin) [reply samples/client/petstore/spring-stubs/src/main/java/org/openapitools/api/PetApi.java:67]: Me neither

**Review by borsch — APPROVED** [review]
@rcardin I have find where that blank line comes from + I've added generation sample with `useResponseEntity=true`. Once pipeline is successful, I'll merge this changes

Thanks

**Comment by rcardin** [comment]
@borsch , thanks a lot for all of your support ❤️
